pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -374,19 +377,28 @@ func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
 		old.stop()
 	}
 
-	go func() {
-		e.run()
-		e.stop()
+	e.run()
+	e.stop()
 
-		m.mutex.Lock()
-		if old, ok := m.endpoints[r.ResourceName]; ok && old == e {
-			glog.V(2).Infof("Delete resource for endpoint %v", e)
-			delete(m.endpoints, r.ResourceName)
+	m.mutex.Lock()
+	if old, ok := m.endpoints[r.ResourceName]; ok && old == e {
+		glog.V(2).Infof("Mark all resources Unhealthy for endpoint %v", e)
+		// NOTE: Intentionally endpoint is not being cleaned up from cache.
```

Comment:
If we leave the endpoint around, one potential problem is for a device plugin that requests preStart, we may keep issuing the grpc call for pods that are already allocated. Is it possible that we remove the endpoint but modify GetCapacity() to not populate deletedResources?

Reply:
Even if we remove endpoint from cache, for a device plugin that requests preStart, it will still keep issuing the grpc call for pods that are already allocated. It is because options device options are stored in the manager. Instead of removing endpoint, we can remove options for a deleted endpoint from manager cache, then PreStart will not be invoked repeatedly in a scenario like you mentioned.

Reply:
Thinking more on this, If device plugin has asked for preStart, then repeated preStart calls seems expected behavior. And PreStart will fail each time. But yeah,  If we remove endpoint from cache, error message will be more intutive, which is about missing endpoint.